`profsea-climate`: Add new Antarctic component

gregrmunday opened the pull request:
I've been working on an Antarctic ice-sheet emulation of ISMIP6 2300 simulations which responds more realistically to overshoot and idealised temperature forcings. 

Addresses #27 

I've added the component here but it still needs to be worked into the main emulator code!

gregrmunday committed:
Added new antarctic component

gregrmunday committed:
Improve performance + add ais_params.nc

gregrmunday commented:
A look at how it behaves over a crazy ramp up ramp way down forcing:

<img width="819" height="653" alt="Screenshot 2025-12-05 at 09 18 58" src="https://github.com/user-attachments/assets/5eaa5256-d23d-45ee-b8e0-69197a316731" />

gregrmunday committed:
Add AIS and other notebooks

gregrmunday commented on profsea/notebooks/ais_fitting.ipynb:
@mo-jenniferweeks I've uploaded the AIS emulator fitting notebook here - I'm still changing some things around in it and it might be a bit messy at the moment!

I'm working on fitting the emulator with the 19 common IS models across all scenarios as opposed to 43 across half of the scenarios to see how it compares!

gregrmunday commented:
WIP comparison between AR5 emulator and ISMIP6 emulator:

<img width="1079" height="473" alt="Screenshot 2025-12-05 at 14 02 51" src="https://github.com/user-attachments/assets/be9f6e87-95e7-4542-9ed8-f2dc34850dbb" />

<img width="1006" height="509" alt="Screenshot 2025-12-05 at 14 02 33" src="https://github.com/user-attachments/assets/61e353f8-4bd7-4032-bd10-58852db1440a" />

Plumes are 66% likely range. At the moment it looks a fair bit less sensitive than the AIS emulator to warming, at least over this timescale - although the ramp up in SLE under higher emissions looks a lot more non-linear.  

Next step is probably to compare this against AR6 and against the other emulators in SLEIP

gregrmunday commented:
Comparing ISMIP6 emulation including only 14 of the contributing models, which submitted simulations over all 10 forcing scenarios.

<img width="1079" height="475" alt="Screenshot 2025-12-05 at 15 14 38" src="https://github.com/user-attachments/assets/46437f4d-1b2f-486e-96e0-c70771010761" />

Seems to just lower the projection a bit, the spread is almost captured but this set of models happens to be biased towards a smaller response to temperature, and have a high proportion of long-timescale responses.

gregrmunday committed:
Updated code

gregrmunday committed:
Dev updates

hmkhatri commented:
It is interesting that ISMIP6 response is much smaller than AR5 (and possibly AR6). I think we should keep both options in the code - (1) standard AR5/AR6  (2) improved ISMIP. So, users can choose the algorithm they prefer.  Of course, we will need references and some disclaimer about different methods and what causes these differences in projected sea level.

gregrmunday commented:
> It is interesting that ISMIP6 response is much smaller than AR5 (and possibly AR6). I think we should keep both options in the code - (1) standard AR5/AR6 (2) improved ISMIP. So, users can choose the algorithm they prefer. Of course, we will need references and some disclaimer about different methods and what causes these differences in projected sea level.

Hmm yes, I'm not totally satisfied with this to be honest. The parameterisation captures the range of the ISMIP6 ensemble well but it does underestimate the median by quite some bit, so I think still a bit of work required to find a better fit to the data!

gregrmunday commented:
> It is interesting that ISMIP6 response is much smaller than AR5 (and possibly AR6). I think we should keep both options in the code - (1) standard AR5/AR6 (2) improved ISMIP. So, users can choose the algorithm they prefer. Of course, we will need references and some disclaimer about different methods and what causes these differences in projected sea level.

But I agree that the user should be able to choose which parameterisation to go with. We can use whichever one we feel most confident in as the default, eventually.

gregrmunday committed:
model update

gregrmunday committed:
Merge branch 'profsea-climate' into gm/add_AIS_emulator

gregrmunday committed:
Resolve merge conflicts

mo-jenniferweeks commented:
@gregrmunday is it possible to compare with the Edwards ISMIP emulation too, even though it is to 2100 only?

gregrmunday commented:
> @gregrmunday is it possible to compare with the Edwards ISMIP emulation too, even though it is to 2100 only?

Good idea, I'll assume the data is available with their paper?

gregrmunday committed:
Merge branch 'profsea-climate' into gm/add_AIS_emulator

gregrmunday committed:
More AIS fitting...

gregrmunday committed:
Update comparison notebook

gregrmunday committed:
Update for latest version

gregrmunday committed:
Mo' updates

gregrmunday committed:
Smol fix

gregrmunday committed:
Optional spatial sampling

gregrmunday committed:
Remove import

gregrmunday committed:
Add params

gregrmunday committed:
Add dim check and reformat

gregrmunday committed:
Fix bugs in probabilistic_wrapper.py

gregrmunday committed:
Remove unnecessary notebooks

gregrmunday commented:
@hmkhatri  this is ready for review! 

Lots of changes throughout this one:

- [x] Optimisation and refactor of the `Global` class
- [x] Added the `ISMIP6Emulator` to the `Global` class
- [x] Updated `probabilistic_wrapper.py` and fixed some bugs 
- [x] Fixed bugs in, improved the performance of, and improved the usability of the `Spatial` class

---

I would like to think about making all the auxiliary data required to run `ProFSea` easily available in some shape or form. One approach would be like how we do it in `SpeedyWeather`, with a [separate repo](https://github.com/SpeedyWeather/SpeedyWeatherAssets) for data (call it something like `ProFSeaAssets`). I'm sure we could compress all the CMIP6 and GRD/GIA patterns small enough to do that. 

In that case we could include some code to download it locally when a `ProFSea` user calls `Spatial()` the first time, then after that it's already local. This could be a good idea as not every `ProFSea` user will want to produce spatial projections, they may just want to quickly make some global ones in which case having to download the extra spatial-related data is a pain.

If we can compress the spatial data *very* small, then we could just include it in the repo, but this could be less ideal.

gregrmunday committed:
Fix source attribute bug

hmkhatri commented:
@gregrmunday Thanks for the pull request. I will review it this week.

hmkhatri commented on profsea/emulator/antarctica.py:
@gregrmunday Could you add some comments about the approach in `antarctica.py` ? This is the main part of the calculation and some notes would be useful

gregrmunday replied:
For sure! Will upload in the next commit

gregrmunday replied:
Added this in the latest commit. I don't think we need to go into excessive detail here as it'll be detailed in the paper.

hmkhatri commented on profsea/emulator/gmslr.py:
```diff
@@ -161,15 +161,14 @@ def __init__(
         self.random_sample = random_sample
         self.T_percentile_95 = T_percentile_95
         self.OHC_percentile_95 = OHC_percentile_95
-        self.cum_emissions_total = cum_emissions_total
```
I suggest retaining this. If we want to use AR5 method for icesheets, then we would need cumulative emissions.
Similar to `glaciermip: bool | int = 2`, could provide option to choose between ISMIP or AR5 method.

gregrmunday replied:
Alright I've kept this in. I agree we should come up with a nice way for users to choose how they want to build the model up with different components...

hmkhatri commented on profsea/emulator/gmslr.py:
```diff
@@ -186,54 +185,66 @@ def __init__(
+        ]
+
+        # Setup AIS emulator, if needed
+        if "antarctica" in self.active_components:
+            wais_path = Path(__file__).parent / "aux_data" / "wais_params.nc"
+            eais_path = Path(__file__).parent / "aux_data" / "eais_params.nc"
+            aispen_path = Path(__file__).parent / "aux_data" / "aispen_params.nc"
+            self.wais_model = AntarcticaISMIP6(wais_path)
```
Is ActarcticaISMIP6 the only method for antractica component or is there an option to run with the older method? Better to retain both, in case one wants to run compare the two.

gregrmunday replied:
It's not the only way, the user can specify `antsmb` and `antdyn`, although maybe this should be simplified with a general `antarctica_ar5` component which models both.

hmkhatri commented on profsea/emulator/gmslr.py:
```diff
@@ -245,126 +256,213 @@ def save_components(self, output_dir: str, scenario_name: str) -> None:
+        self.scenario = scenario
+        self.T_change = np.asarray(T_change)
+        self.OHC_change = np.asarray(OHC_change)
+        self.cum_emissions_total = cum_emissions_total
```
There is no option to supply `cum_emissions_total`. Sugegst keeping this input for now.
https://github.com/MetOffice/ProFSea-tool/blob/dd7ef43282412e0ba7ef0fcd559780646fb83be2/profsea/emulator/gmslr.py#L135-L155

gregrmunday replied:
I've re-added this

hmkhatri commented on profsea/emulator/gmslr.py:
```diff
@@ -376,125 +474,178 @@ def calculate_drivers(self) -> tuple:
+                    ),
+                ]
+            else:
+                raise KeyError("glaciermip must be 1 or 2")
```
It seems `glaciermip = 3` is a valid option. I guess valid options for glaciermip input are `False`, 1, ,2 or 3. Is this correct? 
Even if glaciermip is set to `False`, the else condition kicks in and computes its contribution.
https://github.com/MetOffice/ProFSea-tool/blob/dd7ef43282412e0ba7ef0fcd559780646fb83be2/profsea/emulator/gmslr.py#L600-L615

I am confused about which valid options are available for a user.

gregrmunday replied:
I've removed the `glaciermip3` option - I thought about including results from the latest `glaciermip` but I think this should be tackled in its own PR later down the line.

hmkhatri commented on profsea/emulator/spatial.py:
```diff
@@ -70,11 +71,11 @@ def __init__(
                 "Provide namelist of global projection components. "
                 "Currently allowed components are expansion, antdyn, "
-                "antsmb, glacier, greenland and landwater")
+                "antsmb, wais, eais, glacier, greenland and landwater")
```
Need some info on how antsmb, antdyn are different from wais, eais.

gregrmunday replied:
I've update this error message to include more info!

hmkhatri commented on profsea/emulator/spatial.py:
```diff
@@ -70,11 +71,11 @@ def __init__(
         self.components = {}
-        component_path = Path(components_dir) / f"{scenario}_global.nc"
-        ds_component = xr.load_dataset(component_path)
+        component_path = Path(components_dir)
```
Is this correct? Where is the filename of `global.nc` used?

gregrmunday replied:
You're right - this is poorly named, I think the user should provide the exact path to the file here.

hmkhatri commented on profsea/probabilistic_wrapper.py:
```diff
@@ -101,7 +116,32 @@ def load_magicc_forcing(
 
 
-def run_fair(baseline_start: int, baseline_end: int, scenarios: list) -> tuple[np.ndarray]:
+def load_fair_forcing(
```
Now sure about having `load_fair_forcing` function inside ProFSea. I suggest keeping ProFSea more general rather than tied to any specific climate emulator. 

There is code for running FaIR as well here. I would suggest not to keep this in ProFSea. A user should run FaIR (or any other simple climate model) separately and supply inputs as needed.

gregrmunday replied:
I think this is a fair point (no pun intended) but I do think it's super useful to have a script with FaIR baked in lying around somewhere in the repo, even if users don't need to rely on it all the time. We need to think a bit more carefully about what scripts are useful to provide along with the generally usable `Global` and `Spatial` objects, and where we should keep them.

hmkhatri reviewed (changes requested):
@gregrmunday thanks for the pull request. I have made some suggestions. Could you please look into those?

Also, it would be useful to add an option (wherever possible) to choose from different available methods for calculating a component. It provides more clarity to a user on what exactly is running behind ProFSea. 

See issue #38

gregrmunday committed:
Add comments and reformat

gregrmunday committed:
Smol fix

gregrmunday committed:
Address review comments

hmkhatri reviewed (approved):
It is ready to merge.
